Improve MKS Robin support

robindubreuil opened the pull request:
### Updates on the MKS Robin pins configuration file
Hi everyone and thank you for working on the wonderful piece of software that Marlin is.
I managed to update pins_MKS_ROBIN.h to take advantage of most of the board capabilities.

### Changes
- ~~Support for the 16Kb (2KB) onboard I2C EEPROM:**tested working.**~~
- Support for the 64Kb (8KB) onboard SPI FLASH: **not tested.**
- Support for the ColorUI interface on most MKS ROBIN TFT controllers (FSMC): **tested working only with ILI9328 on MKS Robin TFT v1.1R.**
- SD now uses 4.5MHz SDIO (no card detect pin, needs edits to take adavantage of it on the MKS Robin V2.4): **tested working on MKS Robin v2.2 and v2.3, latest bugfix required.**
- Added some hints on the required hardware mod to use SPI2 for additional devices.
- Cleanup and refactoring.
- Miscellaneous minor tweaks.

This code can still be improved as i do not use the LVGL interface because it lacks UBL and advanced pause feature so things may still be missing for the LVGL interface users. More work can be done to add or fix those features on MKS Robin derivatives, as i don't have those boards i prefer to let others improve their pin configuration file on the basis on my work.
Also, i've seen that a lot of work is being done to refactor the TFT interface (see #19192) and it sounds awesome, but it would need to update this file again.

robindubreuil committed:
Update pins_MKS_ROBIN.h

### Updates on the MKS Robin pins configuration file
Hi everyone and thank you for working on the wonderful piece of software that Marlin is.
I managed to update pins_MKS_ROBIN.h to take advantage of most of the board capabilities.

### Changes
- Support for the 16Kb (2KB) onboard I2C EEPROM: **tested working.**
- Support for the 64Kb (8KB) onboard SPI FLASH: **not tested.**
- Support for the emulated DOGM interface on most MKS ROBIN TFT controllers (FSMC): **tested working only with ILI9328 on MKS Robin TFT v1.1R.**
- SD now uses 4.5MHz SDIO (no card detect pin, needs edits to take adavantage of it on the MKS Robin V2.4): **tested working on MKS Robin v2.2, latest bugfix required.**
- Added some hints on the required hardware mod to use SPI2 for additional devices.
- Cleanup and refactoring.

robindubreuil committed:
Update pins_MKS_ROBIN.h

Hi everyone and thank you for working on the wonderful piece of software that Marlin is.
I managed to update pins_MKS_ROBIN.h to take advantage of most of the board capabilities.

- Support for the 16Kb (2KB) onboard I2C EEPROM: **tested working.**
- Support for the 64Kb (8KB) onboard SPI FLASH: **not tested.**
- Support for the emulated DOGM interface on most MKS ROBIN TFT controllers (FSMC): **tested working only with ILI9328 on MKS Robin TFT v1.1R.**
- SD now uses 4.5MHz SDIO (no card detect pin, needs edits to take adavantage of it on the MKS Robin V2.4): **tested working on MKS Robin v2.2, latest bugfix required.**
- Added some hints on the required hardware mod to use SPI2 for additional devices.
- Cleanup and refactoring.

thisiskeithb commented:
> tested working on MKS Robin v2.2, latest bugfix required

Should some of these changed have been made to the Robin 2 pins instead? https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/pins/stm32f4/pins_MKS_ROBIN2.h

TheCodeExorcist commented:
Hi,
MKS Robin and Robin 2 boards are completely different. The kind of board my PR deals with are the first Robin based on the STM32F**103**ZET6 MCU and it appears that mine is a v2.2. Robin 2 boards are newer and more powerful boards based on the STM32F**407**ZET6 MCU, so apparently none of the changes i made could possibly affect those boards (different HAL, pin mappings, features...).

sjasonsmith commented:
@thisiskeithb that is the naming confusion in mentioned the other day. Robin v2 != Robin2

sjasonsmith commented:
@mattdog01 does this seem related to the SDIO issues you’ve been having with a Robin? @TheCodeExorcist, which PlatformIO environment are you using this with? mks_robin or mks_robin_stm23?

rhapsodyv commented:
> Also, i've seen that a lot of work is being done to refactor the TFT interface (see #19192) and it sounds awesome, but it would need to update this file again.

Sure, one of the idea from the TFT refactoring, is to remove the need of board know about the UI. The board will just setup FSMC / SPI interface, and it's done.
There's another place where the display types are defined, with their specs: driver, resolution, orientation, etc. The user just select the UI and the Display type.

TheCodeExorcist commented:
@sjasonsmith, I built this on PlatformmIO for Code OSS withe the standard mks_robin environment. I suspect that the mks_robin_stm32 environment is kind of a relic from the past as i don't see any practical use case for it nowadays...
However, i forgot to mention that i edited W25Qxx.cpp to be able to use the ENABLED() macro here in pins_MKS_ROBIN.h:
`#define HAS_SPI_FLASH
#if ENABLED(HAS_SPI_FLASH)`
So, you have to change those two lines this way in attempt to be able to compile marlin with a stock W25Qxx.cpp:
`#define HAS_SPI_FLASH    1
#if HAS_SPI_FLASH`

robindubreuil committed:
Update pins_MKS_ROBIN.h

Reverting back changes to make file compilable with a stock W25Qxx.cpp file.

robindubreuil committed:
Update pins_MKS_ROBIN.h

Reverting back changes to make file compilable with a stock W25Qxx.cpp file.

sjasonsmith commented:
mks_robin_stm32 is very new. Different overlapping efforts for TFT support resulted in it being first implemented for that environment, then @rhapsodyv later expanding it to support more environments.

TheCodeExorcist commented:
@sjasonsmith I better understand now where this environment comes from, i had some severe compilation issues with it this is why I pushed some efforts to get a functionnal ColorUI on the standard mks_robin environment, on the basis of what has been done on other boards and by reading some of the Marlin code.

mattdog01 commented:
> @mattdog01 does this seem related to the SDIO issues you’ve been having with a Robin? @TheCodeExorcist, which PlatformIO environment are you using this with? mks_robin or mks_robin_stm23?

@sjasonsmith I suspect it does.  I have the MKS-Robin          v2.3 :)  

FYI I ran a couple hour print last night with the latest bugfix available (last night) and it still had the random print head movements to the ends of the axis and back but with a new one, the extruder tried to drive a what seemed like 1000+mm of filament through the hot end in a couple seconds and of course just shredded the filament and then continued to the end of the print. It didn't reset but it did have issues reading the SD card. Sorry if that is too much info.

TheCodeExorcist commented:
@mattdog01 I compiled my firmware last night too with the latest bugfix available and my custom config file. My printer worked like a charm all night long so you should definitely give it a try. I precised that the latest bugfix is needed only because of this fix recently made for boards without SD detect pin: #19236

mattdog01 commented:
@TheCodeExorcist I'm not a coder and know little about how the files should be / are arranged, so I will need to wait for it to get pulled into bugfix for me to get your fixes.  How long does that normally take? 
I looked at the piece of code you have on the other issue.  Does that just need to be pasted into the pins_mks_robin file? I'm assuming at that location?
Any config setting changes?  Build in mks_robin_stm32?

mattdog01 commented:
@TheCodeExorcist I could't find any of this text in my pins_MKS_ROBIN.h in order the know where to change the code there.  Best if I just look and not touch.  

W25Q64 64Mb (8MB) SPI flash
#define HAS_SPI_FLASH

TheCodeExorcist commented:
@mattdog01, many things have been added over the original pins_MKS_ROBIN.h so the easiest way to go is to simply replace the original file by my modified one. You have to build with the mks_robin environment as i did not achieve to build it with mks_robin_stm32 and as i don't see any advantage of using mks_robin_stm32 over mks_robin i don't think i will try to make it work. Your config should remain the same as long as you use FSMC_GRAPHICAL_TFT (for upscaled Marlin UI) or TFT_320x240 (for Color UI) and if you don't switch to one. If you choose TFT_320x240 to take advantage of Color UI, you may get a non working screen and if that happens, try commenting this line `#define TFT_ILI9328` and decommenting the next driver `//#define TFT_R61505` and so on until you find what driver you TFT uses...

sjasonsmith commented:
> i don't see any advantage of using mks_robin_stm32 over mks_robin

The hope is to eventually move everything from HAL/STM32F1 to HAL/STM32, making mks_robin_stm32 the future. That said, there is no immediate timeline for that migration. PlatformIO has deprecated the Maple framework used by HAL/STM32F1, and from a maintenance perspective there are advantages to only maintaining a single STM32 HAL.

@mattdog01 is having some reliability issues with his SD reads using mks_robin_stm32, so this will be a good data point to determine whether the same problems exist when he is using your version with the mks_robin environment.

sjasonsmith commented on Marlin/src/pins/stm32f1/pins_MKS_ROBIN.h:
```diff
@@ -41,24 +43,22 @@
-#define SERVO1_PIN                          PA1   // XS1 - 6
-#define SERVO2_PIN                          PF9   // XS2 - 5
-#define SERVO3_PIN                          PF8   // XS2 - 6
+#define SERVO0_PIN                          PA1   // XS1 - 6 (BLTOUCH)
+#define SERVO1_PIN                          PC3   // XS1 - 5
+#define SERVO2_PIN                          PF8   // XS2 - 6
+#define SERVO3_PIN                          PF9   // XS2 - 5
```
Is rearranging these likely to break existing users?

Did they perhaps move connectors around between board revisions to provide a dedicated BLTouch connector or something?

robindubreuil replied:
I only rearranged SERVO PINS like this because it seemed like PA1 is the only hardware PWM capable pin according to the datasheet so i decided to use it for BLTOUCH as it should apparently be the preferred servo pin.

sjasonsmith replied:
Currently hardware PWM is never used for Servo output in our STM32 HALs, including for a BLTouch.
That doesn't mean it _couldn't_ be used, it just is not today.

Timer conflicts and shortages have been a battle on some of these devices, so on many boards it isn't possible to find pins capable of hardware PWM at all.

Does this board even have pins marked Servo, or are these just on the AUX connector?
It seems bizarre to declare four servo pins at all if they don't actually exist on the board. I can see defining one specifically for a BLTouch, but even that is questionable when the board doesn't have a labeled position for it.

sjasonsmith replied:
Oh, I see the pins this refers to now.

robindubreuil replied:
In fact, these pins are neither servo nor BLTouch marked. Some Robin variants (not covered by this file) have dedicated BLTouch or servo pins but it is not the case for the original Robin board. There are only some expansion headers (simple 2.54mm dual rank pin headers) with some weird pinouts. I didn't know that Marlin cannot make use of hardware PWM, as it _may_ be implemented one day it should still be a good idea to keep PA1 as the preferred servo pin, what do you think ?

sjasonsmith replied:
I _personally_ prefer to leave them the same, so that if anybody is already using Servo0 it will not stop working.

robindubreuil replied:
I understand, and i think that if there is a way to specify the servo pin used for BLTouch it is the right way to go. If not i think that maybe the feedback of more Robin users like @mattdog01 would make us decide.

thinkyhead replied:
When a servo probe or BLTouch is in use, the `Z_PROBE_SERVO_NR` defines which servo port it uses.

TheCodeExorcist commented:
@sjasonsmith Ok, so with that said every STM32 based board should move to the unified STM32 HAL with no real deadline for now. Is the MKS Robin use of the unified STM32 HAL a standard way of doing it (all boards should do the same way) or something still hacky ?
It is in fact interesting to see if @mattdog01 have the same issues on mks_robin because this new HAL seems to be in use for a few boards only so we have less feedback on running marlin with it. I think this should still be considered unsafe for production, no ?

sjasonsmith commented on Marlin/src/pins/stm32f1/pins_MKS_ROBIN.h:
```diff
@@ -101,117 +101,190 @@
+
+//
+// Power loss detection
+//
 #define POWER_LOSS_PIN                      PA2   // PW_DET
+
+//
+// Power supply control
+//
 #define PS_ON_PIN                           PA3   // PW_OFF
-#define FIL_RUNOUT_PIN                      PF11  // MT_DET
 
-#ifdef ARDUINO_ARCH_STM32F1
-  #define BEEPER_PIN                        PC13
-#else
-  #define BEEPER_PIN                        -1
-#endif
+//
+// Piezzoelectric speaker
+//
+#define BEEPER_PIN                          PC13
```
I wonder why this was `-1` when using the STM32 HAL?

This might introduce a timer conflict for `mks_robin_stm32`, but I'm not too worried about that. If that happens a little extra work will be needed to possibly move the TONE_TIMER, but that would be the right thing to do.

I verified that PC13 isn't attached to any timer pins. The Maple framework will take over any timer attached to BEEPER_PIN if SPEAKER is enabled, which can make enabling it risky for some boards. It seems ok for this board.

robindubreuil replied:
The beeper works fine when compiled with mks_robin on my board, i should try compiling with mks_robin_stm32 to see if everything is still fine but for now i have other issues to solve to make marlin compile on this framework with my config...

jmz52 replied:
>I wonder why this was -1 when using the STM32 HAL?

There was a bug that caused Marlin with HAL STM32 hangs when speaker enabled.
I have not tested this for a while, so I can't tell if this bug still present.

sjasonsmith replied:
If you tested before I upgraded to ststm32 8.0 that might be fixed. I think I remember TONE not working on one timers, but it worked with the newer framework.

rhapsodyv replied:
The only problem in BEEPER_PIN that I know is when SPEAKER is enabled. I'm using the BEEPER in a STM32F4, on HAL STM32 with no problem, in the current version.

sjasonsmith commented on Marlin/src/pins/stm32f1/pins_MKS_ROBIN.h:
```diff
@@ -101,117 +101,190 @@
-  #if HAS_TMC_SW_SERIAL
-    #define TMC_BAUD_RATE                  19200
+  // Hardware serial on Serial 1
+  #define TMC_HARDWARE_SERIAL
+  #if ENABLED(TMC_HARDWARE_SERIAL)
+    #define X_HARDWARE_SERIAL               Serial1
+    #define X2_HARDWARE_SERIAL              Serial1
+    #define Y_HARDWARE_SERIAL               Serial1
+    #define Y2_HARDWARE_SERIAL              Serial1
+    #define Z_HARDWARE_SERIAL               Serial1
+    #define Z2_HARDWARE_SERIAL              Serial1
+    #define E0_HARDWARE_SERIAL              Serial1
+    #define E1_HARDWARE_SERIAL              Serial1
+    #define E2_HARDWARE_SERIAL              Serial1
+    #define E3_HARDWARE_SERIAL              Serial1
+    #define E4_HARDWARE_SERIAL              Serial1
+  #endif
+
+  // Software serial on unused servo pins
+  #define TMC_SOFTWARE_SERIAL
```
You are enabling both TMC_HARDWARE_SERIAL and TMC_SOFTWARE_SERIAL.

robindubreuil replied:
My bad, i'll fix it in no time.

sjasonsmith commented:
@TheCodeExorcist I only just discovered that mks_robin_stm32 environment. It slipped in with a TFT PR, so I didn't notice it happened (I tend to ignore display-related work).

I haven't really looked at it close enough to have an opinion on it. I would not assume it is perfect, especially with one of the few early users reporting SD reliability issues.

Nobody has really formulated a plan for moving things over. I would prefer an approach where we get one or two boards really stable and well-defined, which then serves as a template for the rest. The real test will be moving the SKR E3 Mini over, to see if a week can go by without a ton of new issues reported!

robindubreuil committed:
Update pins_MKS_ROBIN.h

commenting TMC SOFTWARE SERIAL by default

robindubreuil committed:
Update pins_MKS_ROBIN.h

commenting TMC SOFTWARE SERIAL by default

TheCodeExorcist commented:
> Nobody has really formulated a plan for moving things over. I would prefer an approach where we get one or two boards really stable and well-defined, which then serves as a template for the rest. The real test will be moving the SKR E3 Mini over, to see if a week can go by without a ton of new issues reported!
This would be indeed a quite fireproof test, i'll try to see what i can do on the Robin side.

sjasonsmith commented:
In the meantime, it is fine to continue fixing things in the STM32F1 versions, I just wouldn't recommend large development investments in it.

mattdog01 commented:
@TheCodeExorcist  So far so good.  I went to your repo copied and pasted the contents of pins_MKS_ROBIN.h to my pins_MKS_ROBIN.h .  It compiled just fine, The new UI looks good.  SD card was ready to be used, no need to use the custom commands to get it working.
@sjasonsmith I think I will need to specify a different set of servo pins and move my servo wires to that location because my servo for bed leveling doesn't work anymore.

mattdog01 commented:
@sjasonsmith No, way easier than that.  Just move the connector to the next servo connection on the board. Works great.

thinkyhead committed:
Apply auto-format

thinkyhead committed:
Apply auto-format

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19333

thinkyhead commented on Marlin/src/pins/stm32f1/pins_MKS_ROBIN.h:
```diff
@@ -41,24 +43,22 @@
-  #define SDCARD_EEPROM_EMULATION
+#if NO_EEPROM_SELECTED
+  #define IIC_EEPROM_SDA                    PB7
+  #define IIC_EEPROM_SCL                    PB6
```
Marlin only needs these two pin definitions when using the BL24C16 chip included on Creality 4.x boards. If this board has a regular wired (I2C or SPI) EEPROM, the better choice would be `I2C_EEPROM` or `SPI_EEPROM`. If it doesn't have one of these then Flash-emulated EEPROM is preferred so that the SD card can be removed without any consequence.

robindubreuil replied:
The original MKS Robin board have two external storage devices:
- The AT24C16N-10SI-2.4 16Kb (2KB) I2C EEPROM (but some variants seems to have unknown chinese clones).
- The W25Q64BVSSI 64Kb (8KB) SPI FLASH.

My goal was to make use of the I2C EEPROM as i don't know what can marlin do with a SPÏ flash except using it as an emulated EEPROM. I tried to simply use I2C_EEPROM and it compiled just fine, but marlin was frozen on the boot screen as if no EEPROM was detected. After hours of searching, i tried to use the Creality way of defining the I2C EEPROM and it worked just fine. I don't know what exactly this changes in the behavior of marlin, according to you it seems to be a way to communicate with EEPROM chips that have been weirdly wired and maybe this is the case for the Robin v2.2 I have as it works just fine with this (and it also seems to be the case for @mattdog01 Robin v2.3).

I also added a pin definition for the W25Q64BVSSI SPI flash and it compiles. I don't know if it works as my marlin don't use the SPI flash chip. If anyone knows how to make an interesting use of this chip (such as deporting ColorUI assets to it for example...) i'd be really interested.

sjasonsmith replied:
I think @rhapsodyv was doing SPI flash work, although in his case it was flash on a display PCB and not on the motherboard itself.

rhapsodyv replied:
We use spi flash to store LVGL icons. We use about 1mb to 2mb.

mattdog01 commented:
@TheCodeExorcist @sjasonsmith Well that was perfect.  No random print head movements, just nice clean printing.  Very stable. that is a winner.  Thanks so much guys for all your work.  I love the new GUI, that it is stable and easy to configure.  I recommend that some additional text be added to the config.h file to explain how to enable the different display styles.  So nice.

sjasonsmith commented:
@mattdog01 can you make a comment about your experience with this PR on your open issue. We're going to have to try to figure out why it is failing with the other HAL, since as I said earlier, it is hopefully going to be the one everyone uses in the future.

mattdog01 commented:
@sjasonsmith I'm just doing that now and connecting it to this PR.  I hope that helps.  I will leave the other issue open though, right?

sjasonsmith commented:
@mattdog01 yes, leave the other one open. Somebody may ask you to help try out a fix if the source of those problems is found!

That also was a good test for this pull request, your testing helps demonstrate the value it is providing.

thinkyhead committed:
0xFF

TheCodeExorcist commented:
All right,
I had REALLY ANNOYING issues with the original bootloader of my v2.2 Robin board (see [this post](https://github.com/makerbase-mks/MKS-Robin/issues/187), don't hesitate to show your support on it if you are the proud owner of a Robin board). I managed to flash the bootloader of a v2.4 Robin board and all seems better now (the absence of SD card detect pin does not seem to cause any issue). I managed to compile the latest bugfix using the mks_robin_stm32 environment along with my new pin file BUT the SPI FLASH section must be removed as the W25Qxx.cpp way of handling SPI seems not compliant with this newer HAL. I tried to see what i can do to make this file compliant with the generic HAL but i think there are some people way more aware of the file and the newer HAL over here that can achieve better work and thus quicker. So, marlin compiles under the mks_robin_stm32 environment with the SPI FLASH section removed from the pin file however marlin does not start: after flashing or boot the bootloader just says "complete" and nothing happens. I tried to flash marlin without bootloader through ST-LINK V2, the steptsicks are enabled and nothing happens (no LED flashing, no display at all). I'll continue testing further, if anyone has ideas they are welcome.

rhapsodyv commented:
I fixed W25Qxx.cpp, and everything else related to SPI / TFT in HAL STM32. I will send a PR soon.

rhapsodyv commented:
I found and fixed the problem with the buzzer :-)

dandantsui commented:
I have a MKS Robin v2.4 board if you need me to try out anything? Noticed about the SDIO pin?

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19333

thinkyhead committed:
Use 0xFF (not 'ff') for byte transfer

thinkyhead committed:
Move TFT_320x240 pin entries

thinkyhead committed:
Move TFT_320x240 pin entries

thinkyhead committed:
Catch undefined SPI_DEVICE

thinkyhead committed:
Catch undefined SPI_DEVICE

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19333

thinkyhead committed:
Don't fall back to BL24CXX

thinkyhead committed:
Don't fall back to BL24CXX

thinkyhead committed:
Allow touch calibration override

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19333

thinkyhead commented:
At this time the build is failing because `HAS_SPI_FLASH` is defined, but `SPI_DEVICE` is not provided.

rhapsodyv commented:
> At this time the build is failing because `HAS_SPI_FLASH` is defined, but `SPI_DEVICE` is not provided.

The W25QXXFlash uses methods from SPIClass that only available in STM32F1. 

As I said before in this thread, I have done it's port for STM32, and made all UI working. But it's depend on #19192 .

thinkyhead committed:
Merge branch 'bugfix-2.0.x' into patch-1

thinkyhead committed:
Merge branch 'bugfix-2.0.x' into patch-1

thinkyhead committed:
Update pins_MKS_ROBIN.h

thinkyhead committed:
Update pins_MKS_ROBIN.h

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19333

thinkyhead committed:
Disable TFT_ILI9328 in the pins file

thinkyhead committed:
Disable TFT_ILI9328 in the pins file

thinkyhead committed:
Fix a TFT define

thinkyhead committed:
Fix a TFT define

thinkyhead commented:
This is just about passing. The only thing missing now is to make sure that `SPI_DEVICE` is properly defined for the `mks_robin_stm32` target. At the moment it is not being defined anywhere in HAL/STM32.

sjasonsmith commented:
@TheCodeExorcist are you still working on this? Do you need help with any of the remaining issues?

sjasonsmith commented on Marlin/src/pins/stm32f1/pins_MKS_ROBIN.h:
```diff
@@ -144,71 +187,144 @@
+
+  // Software serial on unused servo pins
+  //#define TMC_SOFTWARE_SERIAL
+  #if ENABLED(TMC_SOFTWARE_SERIAL)
```
You need to set TMC_BAUD_RATE to 19200 in the Software Serial case. The default baud rate will not work, at least not reliably.

sjasonsmith commented on Marlin/src/pins/stm32f1/pins_MKS_ROBIN.h:
```diff
@@ -144,71 +187,144 @@
-    #define TMC_BAUD_RATE                  19200
+  // Hardware serial on Serial 1
+  #define TMC_HARDWARE_SERIAL
+  #if ENABLED(TMC_HARDWARE_SERIAL)
+    #define X_HARDWARE_SERIAL            Serial1
```
In HAL/STM32F1 these now have to be `MSerial1` instead of `Serial1`. I added a compile-time check to make sure the `Serial` names are not used, because their interrupt routine is able to hang the board.

That is complicated a little by this pins file being used by both STM32 and STM32F1 HALs...
_Maybe_ MSerial1 is going to be fine in HAL/STM32, but I haven't tried it. I think we had made efforts to make sure the emergency parser wasn't running on it in HAL/STM32F1, but that probably wasn't done in STM32.

Is it practical for the user to use Serial1 on this board? Using that as the default is appropriate?

sjasonsmith commented on Marlin/src/pins/stm32f1/pins_MKS_ROBIN.h:
```diff
@@ -144,71 +187,144 @@
+//
 #if HAS_TMC_UART
+
+  // Reduce baud rate for software serial reliability
+  #if HAS_TMC_SW_SERIAL
+    #define TMC_BAUD_RATE                  19200
+  #endif
```
I commented about baud rate lower in the file.

I don't think this is going to work, because `HAS_TMC_SW_SERIAL` depends on the pin definitions which come later in the file!

rhapsodyv commented:
Now this board will work on STM32. I just pushed the code needed to work with W25QXXFlash class: #19989

sjasonsmith commented:
@TheCodeExorcist are you still around, or is this abandoned?

sjasonsmith commented:
@rhapsodyv this is the board you use, right? Do the changes in this PR help you? It seems someone else may have to take these changes over and finish them.

rhapsodyv commented:
@sjasonsmith it not this board I have. But is similar and my work impact on that board too.

Someone needs to merge current marlin back to this branch (I don't have permission). It will fix the W25Qxx error, allowing this PR to be merged.

sjasonsmith commented:
@rhapsodyv I thought I had this board, but it seems I only have a Robin2 and not a Robin V2.

sjasonsmith commented:
@caza5110 you "approved" these changes. Do you have this hardware, and are able to verify they work?

dandantsui commented:
Is this on the bugfix branch yet? I can test if needed. I have a Robin v2.4

sjasonsmith commented:
@dandantsui it is not in bugfix yet. The original submitter hasn't responded for some time. I'm querying to see if we can find a way to finish it up so it can be merged.

TheCodeExorcist commented:
Hi everyone,
Sorry for my absence, i run my own company in France and because of the lockdown I have not much time to spend on this project.
I discovered that what i thought was a fix concerning EEPROM was in reality absolutely not working (it silently falls back to SD EEPROM emulation on the maple based HAL and just crashes on the common STM32 HAL). I'm since a while trying to find a way to make the EEPROM work but with no luck. It seem that you fixed the W25QXX issue so the pin file should produce a working result with the STM32 environment, I will try various configs with the new HAL (this weekend) to see if i could get the hardware EEPROM finally working and i encourage anyone having this board and time to spend to do so.

@sjasonsmith, it seems that you found many caveats regarding the TMC serial communication. As i run my robin with THB6128 (S6121 v1.1) drivers and i don't have any printer with TMC drivers (I had various issues with >1.5A RMS rated trinamic drivers running with >1.5A RMS stepper motors so I tend not to buy them), I don't have any hardware to test or improve the parts that concerns them. According to [the schematics of my v2.2 robin board](https://raw.githubusercontent.com/makerbase-mks/MKS-Robin/master/MKS%20Robin/Hardware/MKS%20Robin%20V2.2_001%20SCH.pdf) , the serial0 interface is used to communicate with the ESP12-S WiFi bridge and the serial1 interface is used for USB communication through CH340G. No other serial interface seems available. I think using hardware serial0 is the easiest method for those willing to use hardware serial on the board because it is easy to de-solder the ESP12-S module or solder cables on it's pins (moreover, some robin boards seems to have the ESP12-S module attached through pin headers).

sjasonsmith commented:
@TheCodeExorcist thanks for the update! I recommend you update the portions you know are working. The rest can remain unchanged for someone else to update and test. That would let your changes be submitted without waiting for everything to be perfect.

If you have portions that you think will work but are not sure, perhaps they could be commented out, with a description that they need verified?

robindubreuil committed:
Update pins_MKS_ROBIN.h

Adding a comment for MKS Robin v2.4 owners and changing defaults for an improved compatibility with the newer HAL.

robindubreuil committed:
Update pins_MKS_ROBIN.h

Adding a comment for MKS Robin v2.4 owners and changing defaults for an improved compatibility with the newer HAL.

robindubreuil committed:
Update pins_MKS_ROBIN.h

Correction of a copy-paste mistake.

robindubreuil committed:
Update pins_MKS_ROBIN.h

Correction of a copy-paste mistake.

robindubreuil committed:
Update pins_MKS_ROBIN.h

Most people knows about the environment but not the HALs...

sjasonsmith commented:
@TheCodeExorcist are you still working on this, or are you done with your changes? There's clearly a merge conflict to resolve, but I'm not sure if we are waiting for you, or you are waiting for us 🙂

TheCodeExorcist commented:
Hi,
I don't understand clearly what should be changed to resolve what appears to be a conflict. On my side i'm still working on the EEPROM support, but i lack time. Even so, the file should be working as is (EEPROM config reverted).

sjasonsmith committed:
Align better with upstream changes

sjasonsmith committed:
Rearrange and cleanup

sjasonsmith committed:
Merge remote-tracking branch 'TheCodeExorcist/patch-1' into PRR/19333_patch-1

# Conflicts:
#	Marlin/src/pins/stm32f1/pins_MKS_ROBIN.h

sjasonsmith commented:
I have rebased this on the current state of bugfix-2.0.x. I guarantee I have made mistakes. @rhapsodyv is going to review the code, the TFT and SD definitions in particular.

I did not force push this. I rebased, then merged the remote branch into my rebase. This means my rebase is reversible if I made a horrible mess of things.

sjasonsmith committed:
Fix failed tests

sjasonsmith commented:
@rhapsodyv I don't know how to configure TFT displays. I left the various display sections in the file, for you to review. I have no idea whether these are still necessary or are correct.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19333

sjasonsmith committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into PRR/19333_patch-1

# Conflicts:
#	Marlin/src/pins/stm32f1/pins_MKS_ROBIN.h

sjasonsmith committed:
Remove ONBOARD SD from Robin

sjasonsmith commented:
I resolved merge conflicts on this again. I think it can only move forward if @rhapsodyv or @TheCodeExorcist  respond and provide feedback. I had to do a lot of "educating guessing" when I updated this 20 days ago.

rhapsodyv committed:
outdated spi configs

rhapsodyv committed:
outdated TFT changes

rhapsodyv commented:
@mattdog01 are you using the pins from here? or from bugfix?

sjasonsmith commented on Marlin/src/libs/W25Qxx.cpp:
```diff
@@ -53,19 +53,9 @@ void W25QXXFlash::init(uint8_t spiRate) {
 
   OUT_WRITE(SPI_FLASH_CS_PIN, HIGH);
 
-  /**
-   * STM32F1 APB2 = 72MHz, APB1 = 36MHz, max SPI speed of this MCU if 18Mhz
-   * STM32F1 has 3 SPI ports, SPI1 in APB2, SPI2/SPI3 in APB1
-   * so the minimum prescale of SPI1 is DIV4, SPI2/SPI3 is DIV2
-   */
-  #if SPI_DEVICE == 1
-    #define SPI_CLOCK_MAX SPI_CLOCK_DIV4
-  #else
-    #define SPI_CLOCK_MAX SPI_CLOCK_DIV2
-  #endif
   uint8_t clock;
   switch (spiRate) {
-    case SPI_FULL_SPEED:    clock = SPI_CLOCK_MAX;  break;
```
This change seems like it does not belong with this PR. @rhapsodyv do you know anything about this?

rhapsodyv replied:
Sure. In fact we can just keep as it is today. I'm changing those spi calls anyway.

thinkyhead replied:
Keep it as it is _here_ today? Or keep it as it is _in bugfix_ today, before these changes?

mattdog01 commented:
> @mattdog01 are you using the pins from here? or from bugfix?

@rhapsodyv I was using the bugfix files without any changes except in the configuration files.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/19333

thinkyhead committed:
Revert W25Qxx.cpp changes

ETE-Design commented:
@TheCodeExorcist You write "SD now uses 4.5MHz SDIO (no card detect pin, needs edits to take adavantage of it on the MKS Robin V2.4): tested working on MKS Robin v2.2 and v2.3, latest bugfix required." I have an MKS Robin Lite who dosen't seem to have card pin, what needed to be changed to use it with that board if possible?

TheCodeExorcist commented:
Hi,
The MKS Robin Lite uses a cheaper variant of the STM32F1 microcontroller which has less IO so the pinout probably differs from the original MKS Robin board in many ways. You should try to understand how this pin file is written and explore the open source MKS schematics that concerns your board to write a relevant pin file for the Robin lite, if nobody did it already.
Have fun.

7 janv. 2021 8:00:15 PM ETE-Design <notifications@github.com>:

> @TheCodeExorcist[https://github.com/TheCodeExorcist] You write "SD now uses 4.5MHz SDIO (no card detect pin, needs edits to take adavantage of it on the MKS Robin V2.4): tested working on MKS Robin v2.2 and v2.3, latest bugfix required." I have an MKS Robin Lite who dosen't seem to have card pin, what needed to be changed to use it with that board if possible?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub[https://github.com/MarlinFirmware/Marlin/pull/19333#issuecomment-756314241], or unsubscribe[https://github.com/notifications/unsubscribe-auth/APAXBSEANSQISOLYYFU5RDTSYYAD3ANCNFSM4REWC2IQ].
>  [data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEgAAABICAYAAABV7bNHAAAABHNCSVQICAgIfAhkiAAAACtJREFUeJztwQENAAAAwqD3T20ON6AAAAAAAAAAAAAAAAAAAAAAAAAAAD4MUUgAARy2AfAAAAAASUVORK5CYII=###24x24:true###][Image de pistage][https://github.com/notifications/beacon/APAXBSCIZXBLVQRCHWD3DILSYYAD3A5CNFSM4REWC2I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFUKHBAI.gif]
>

ETE-Design commented:
@TheCodeExorcist Thanks for taking your time to answer me thougt the PR is closed... Mabye you can help me a little?
How to find/know #define SDIO_CLOCK and #define SDIO_READ_RETRIES
Also what is WIFI_IO0_PIN for?

TheCodeExorcist commented:
The SDIO values should work flawlessly on any STM32F1 based board, just use the same for yours.
The WIFI_IO0_PIN is supposedly useful in an effort to make the ESP12 module (with MKS proprietary formware) work with marlin. I never used nor tested any WiFi functionnality on my board so many things may be needed to make things work properly when it comes to WiFi.

7 janv. 2021 11:58:53 PM ETE-Design <notifications@github.com>:

> @TheCodeExorcist[https://github.com/TheCodeExorcist] Thanks for taking your time to answer me thougt the PR is closed... Mabye you can help me a little?
> How to find/know #define SDIO_CLOCK and #define SDIO_READ_RETRIES
> Also what is WIFI_IO0_PIN for?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub[https://github.com/MarlinFirmware/Marlin/pull/19333#issuecomment-756437768], or unsubscribe[https://github.com/notifications/unsubscribe-auth/APAXBSFY5VNQIAV7ZAUWVNDSYY4CXANCNFSM4REWC2IQ].
>  [data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEgAAABICAYAAABV7bNHAAAABHNCSVQICAgIfAhkiAAAACtJREFUeJztwQENAAAAwqD3T20ON6AAAAAAAAAAAAAAAAAAAAAAAAAAAD4MUUgAARy2AfAAAAAASUVORK5CYII=###24x24:true###][Image de pistage][https://github.com/notifications/beacon/APAXBSFVPJ2AGKLYQHYKSZ3SYY4CXA5CNFSM4REWC2I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFULFGCA.gif]
>
